Modules/posixmodule.c

```diff
@@ -10035,6 +10040,205 @@ os_times_impl(PyObject *module)
 #endif /* HAVE_TIMES */
 
 
+#if defined(HAVE_TIMERFD_API)
+#define ONE_SECOND_IN_NS (1000 * 1000 * 1000)
+#define EXTRACT_NSEC(value)  (long)( ( (double)(value) - (time_t)(value) ) * 1e9)
+
+static PyObject *
+build_itimerspec(const struct itimerspec* curr_value)
+{
+    double _interval = (double)(curr_value->it_interval.tv_sec) +
+                       (double)curr_value->it_interval.tv_nsec * 1e-9;
+    PyObject *interval = PyFloat_FromDouble(_interval);
+    if (interval == NULL) {
+        return NULL;
+    }
+    double _value = (double)(curr_value->it_value.tv_sec) +
+                    (double)curr_value->it_value.tv_nsec * 1e-9;
+    PyObject *value = PyFloat_FromDouble(_value);
+    if (value == NULL) {
+        Py_DECREF(interval);
+        return NULL;
+    }
+    PyObject *tuple = PyTuple_Pack(2, interval, value);
+    Py_DECREF(interval);
+    Py_DECREF(value);
+    return tuple;
+}
+
+static PyObject *
+build_itimerspec_ns(const struct itimerspec* curr_value)
+{
+    return PyTuple_Pack(
+        2,
+        PyLong_FromLongLong((long long)curr_value->it_interval.tv_sec * ONE_SECOND_IN_NS + (long long)curr_value->it_interval.tv_nsec),
+        PyLong_FromLongLong((long long)curr_value->it_value.tv_sec * ONE_SECOND_IN_NS + (long long)curr_value->it_value.tv_nsec)
+    );
+}
+
+/*[clinic input]
+os.timerfd_create
+
+    clockid: int
+        CLOCK_REALTIME or CLOCK_MONOTONIC. CLOCK_REALTIME.
+    flags: int
+        0 or  a bit mask of TFD_NONBLOCK or TFD_CLOEXEC.
+    /
+
+Create and return a timerfd notification file descriptor.
+[clinic start generated code]*/
+
+static PyObject *
+os_timerfd_create_impl(PyObject *module, int clockid, int flags)
+/*[clinic end generated code: output=1caae80fb168004a input=639b0f9daa7ce12e]*/
+
+{
+    int fd;
+    Py_BEGIN_ALLOW_THREADS
+    fd = timerfd_create(clockid, flags);
+    Py_END_ALLOW_THREADS
+    if (fd == -1) {
+        return PyErr_SetFromErrno(PyExc_OSError);
+    }
+    return PyLong_FromLong(fd);
+}
+
+
+/*[clinic input]
+os.timerfd_gettime
+
+    fd: fildes
+        a file descriptor for timer.
+    /
+
+Return timerfd value in seconds.
+[clinic start generated code]*/
+
+static PyObject *
+os_timerfd_gettime_impl(PyObject *module, int fd)
+/*[clinic end generated code: output=ec5a94a66cfe6ab4 input=e8a736fe29b08289]*/
+{
+    struct itimerspec curr_value;
+    int result;
+    Py_BEGIN_ALLOW_THREADS
+    result = timerfd_gettime(fd, &curr_value);
+    Py_END_ALLOW_THREADS
+    if (result == -1) {
+        return PyErr_SetFromErrno(PyExc_OSError);
+    }
+    return build_itimerspec(&curr_value);
+}
+
+/*[clinic input]
+os.timerfd_settime
+
+    fd: fildes
+        a file descriptor for timer.
+    flags: int
+        0 or  a bit mask of TFD_TIMER_ABSTIME or TFD_TIMER_CANCEL_ON_SET.
```

Comment:
I think *flags* should default to 0, and perhaps should also be keyword-only. At the least, it should be positional-or-keyword.